### ISSUE TITLE
v3.0.0 MacOS  - Rendering is out of whack

### ISSUE THREAD

**Issue by carlitorendon** [hluk/CopyQ#691]
Hello CopyQ team, as I have to travel between a Windows and MacOS world, I encountered that for the current release for MacOS the rendering is out of what as you keep adding elements. See below:

![image](https://cloud.githubusercontent.com/assets/14046845/25095001/452d4eec-235f-11e7-94f7-b40a572a77eb.png)

Notice those empty rows; they actually have contents, that appears only when selected. For this particular condition, there's nothing of interest in the logs. Also didn't follow any particular process; just started using it, and the condition manifested as I kept using it.

**Comment by hluk** [comment]
I cannot reproduce this on OS X 10.11.6.

Does this happen only occasionally when application window is opened?

You can test [this pre-release](https://github.com/hluk/CopyQ/releases/tag/osx-test) with some rendering issues fixed.

**Comment by carlitorendon** [comment]
@hluk will do and report back.

**Comment by carlitorendon** [comment]
@hluk I ran the pre-release to no avail (see below):

![image](https://cloud.githubusercontent.com/assets/14046845/25159438/66d63dc6-2475-11e7-8a98-d5a9529bee89.png)

I have heavily customized my keyboard shortcuts, so for a moment I thought that this had something to do with it; I reproduce the test case by using the mouse and the toolbar only, so it was a red herring; has nothing to do with keyboard bindings.

I don't know what else could be special on my side. Just using MacOS Sierra. As I pointed before, the logs show nothing interesting. So let me know if there's something else I can provide.

**Comment by hluk** [comment]
Can you test it with [updated pre-release](https://github.com/hluk/CopyQ/releases/tag/osx-test)? (Unfortunately, I still cannot reproduce it so I'll just try to refactor/optimize the rendering code.)

BTW, does the first item contain "test 4" and the second one "test 3"? Are there just 6 items (0 to 5)? This could mean that item sizes are not updated correctly.

**Comment by carlitorendon** [comment]
Hello @hluk, definitely can test the new binary. To answer your question: 

1. I cleared my flow list
2. I started creating items in order: test1, saving it, creating the next one test2, saving it, etc.

Question:
Would it be useful if I provide you my config files? (see below)

![image](https://cloud.githubusercontent.com/assets/14046845/25184263/90eca3a6-24df-11e7-9c34-163bb15ab323.png)

**Comment by carlitorendon** [comment]
@hluk I'm afraid the latest snapshot you gave me, has no change in behavior. I wonder what up.

**Comment by hluk** [comment]
Hmm, I just tested this on Sierra and still see no issue.

Sharing config file may help - the config file should be in "$HOME/Library/Preferences/com.copyq.copyq.plist" (or same file in the portable folder).

You can also try setting environment variable `COPYQ_LOG_LEVEL=DEBUG` before launching the app so the log is more verbose.

```bash
export COPYQ_LOG_LEVEL=DEBUG
copyq
```

**Comment by carlitorendon** [comment]
Hello @hluk , below is the text for the logs with the verbosity set to DEBUG. I'll IM you  the config and the tab info.

`# ./copyq
Note: Started Background activity for: CopyQ Server
Note: Ended activity
DEBUG: Server "/Users/carlosluis/.config/copyq/.copyq_s" started.
DEBUG: Found plugins in application bundle
Note: Loading plugin: /Applications/CopyQ.app/Contents/Plugins/copyq/libitemdata.dylib
Note: Loading plugin: /Applications/CopyQ.app/Contents/Plugins/copyq/libitemencrypted.dylib
Note: Loading plugin: /Applications/CopyQ.app/Contents/Plugins/copyq/libitemfakevim.dylib
Note: Loading plugin: /Applications/CopyQ.app/Contents/Plugins/copyq/libitemimage.dylib
Note: Loading plugin: /Applications/CopyQ.app/Contents/Plugins/copyq/libitemnotes.dylib
Note: Loading plugin: /Applications/CopyQ.app/Contents/Plugins/copyq/libitempinned.dylib
Note: Loading plugin: /Applications/CopyQ.app/Contents/Plugins/copyq/libitemsync.dylib
Note: Loading plugin: /Applications/CopyQ.app/Contents/Plugins/copyq/libitemtags.dylib
Note: Loading plugin: /Applications/CopyQ.app/Contents/Plugins/copyq/libitemtext.dylib
DEBUG: Loading configuration
DEBUG: Tab "&clipboard": Loading items
DEBUG: Tab "&clipboard": 5 items loaded
DEBUG: Created platform window for non-copyq
DEBUG: Configuration loaded
DEBUG: Starting monitor
DEBUG: Remote process: Starting new remote process "/Applications/CopyQ.app/Contents/MacOS/copyq -s  monitor /Users/carlosluis/.config/copyq/.copyq_m".
DEBUG: Remote process: Started.
DEBUG: Configuring monitor
DEBUG: Executing: copyq eval --  if (monitoring()) {updateTitle('')}
DEBUG: Message received: CommandArguments
DEBUG: Client-6595:
DEBUG:     if (monitoring()) {updateTitle('')}
DEBUG: Updating window title
DEBUG: Created platform window for non-copyq
DEBUG: DONE
DEBUG: Executing: copyq eval --  if (monitoring()) {updateFirst()}
DEBUG: Message received: CommandArguments
DEBUG: Client-6596:
DEBUG:     if (monitoring()) {updateFirst()}
DEBUG: New item: Adding
DEBUG: DONE
DEBUG: Executing: copyq eval --  if (monitoring()) {updateTitle()}
DEBUG: Message received: CommandArguments
DEBUG: Client-6597:
DEBUG:     if (monitoring()) {updateTitle()}
DEBUG: Updating window title
DEBUG: Created platform window for non-copyq
DEBUG: DONE
2017-04-21 09:59:15.182 copyq[6593:1348743] Bartender: Loaded BartenderHelperTwoOneThree
DEBUG: Executing: copyq eval --  toggle()
DEBUG: Message received: CommandArguments
DEBUG: Client-6598:
DEBUG:     toggle()
DEBUG: Created platform window for non-copyq
DEBUG: Become normal app
DEBUG: Created platform window for copyq
DEBUG: Raise CopyQ
DEBUG: DONE
DEBUG: Tab "flow": Loading items
DEBUG: Tab "flow": 4 items loaded
DEBUG: Created platform window for non-copyq
DEBUG: Tab "flow": Saving 1 items
DEBUG: Tab "flow": Items saved
DEBUG: Tab "&clipboard": Saving 6 items
DEBUG: Tab "&clipboard": Items saved
DEBUG: Tab "flow": Saving 2 items
DEBUG: Tab "flow": Items saved
DEBUG: Tab "flow": Saving 3 items
DEBUG: Tab "flow": Items saved
DEBUG: Tab "flow": Saving 4 items
DEBUG: Tab "flow": Items saved
DEBUG: Tab "flow": Saving 5 items
DEBUG: Tab "flow": Items saved
DEBUG: Tab "flow": Saving 6 items
DEBUG: Tab "flow": Items saved
DEBUG: Become background app
DEBUG: Executing: copyq eval --  if (monitoring()) {updateTitle('')}
DEBUG: Message received: CommandArguments
DEBUG: Client-6599:
DEBUG:     if (monitoring()) {updateTitle('')}
DEBUG: Updating window title
DEBUG: Created platform window for non-copyq
DEBUG: DONE
DEBUG: Executing: copyq eval --  if (monitoring()) {updateFirst()}
DEBUG: Message received: CommandArguments
DEBUG: Client-6600:
DEBUG:     if (monitoring()) {updateFirst()}
DEBUG: New item: Adding
DEBUG: DONE
DEBUG: Executing: copyq eval --  if (monitoring()) {updateTitle()}
DEBUG: Message received: CommandArguments
DEBUG: Client-6601:
DEBUG:     if (monitoring()) {updateTitle()}
DEBUG: Updating window title
DEBUG: Created platform window for non-copyq
DEBUG: DONE`

**Comment by hluk** [comment]
Unfortunately, even with the configuration you provided, I still have no luck reproducing the issue.

Some system settings or application can cause this. Any ideas what it could be? For example, do you use any additional app that improves power consumption or rendering?

**Comment by carlitorendon** [comment]
@hluk I do have a TON of stuff. I'm a software developer as well (albeit a backend one; don't do user interaction programs), so I have a lot of stuff that helps me become a keyboard ninja (or at least allows me to believe I'm on a path to). 

So you bring a very good point, the root cause may lay there! This may become a long term wish list on my side. We'll see if I can create a brand new MacOS VM and by process of elimination see which app interaction is offending CopyQ. It's likely not going to come too fast (if at all) because most of my time I spent it on windows and there CopyQ works like a champ. We'll see.

**Comment by hluk** [comment]
I had some issues with dependencies for OS X on Travis CI -- so I had to use some new packages and updated the [pre-release](https://github.com/hluk/CopyQ/releases/tag/osx-test).

If it won't help I'll try to revert one bigger commit in 3.0.0 which fixed some window behavior on OS X. That could be the culprit.

**Comment by carlitorendon** [comment]
Hello @hluk will give it a try and report back

**Comment by carlitorendon** [comment]
Hello @hluk it doesn't appear to change the behavior, see below:

![image](https://cloud.githubusercontent.com/assets/14046845/25551993/7ef7f254-2c4b-11e7-9e55-03f60d6aec03.png)

**Comment by hluk** [comment]
@carlitorendon I reverted the commit I mentioned in the last comment. Can you test the new pre-release again? Thanks.

**Comment by hluk** [comment]
I finally managed to reproduce this issue with:

```bash
QT_AUTO_SCREEN_SCALE_FACTOR=1 copyq
```

and

```bash
QT_SCALE_FACTOR=2 copyq
```

Do you use any of `QT_*` environment variables?

Anyway, I'll try to fix that.

**Comment by hluk** [comment]
Hopefully the new [pre-release](https://github.com/hluk/CopyQ/releases/tag/osx-test) fixes this.

**Comment by carlitorendon** [comment]
Great to hear that @hluk, will try it and share findings

**Comment by carlitorendon** [comment]
@hluk Excellent! It works now. Regarding the QT* environment variables I haven't manually changed them. Also when I pipped the output of SET to grep for QT everything came back empty.

The only think I can think of is maybe because I have used other QT based programs (such as wireshark and gimp) and maybe that had some effect on QT?

Don't know but definitely I'm happy (and thankful) for this fix.

![image](https://cloud.githubusercontent.com/assets/14046845/25571960/0a8ef5f8-2df3-11e7-92f3-54c1e530b07c.png)
